File: src/main/java/dev/rvbsm/fsit/api/event/UpdatePoseCallback.java
package dev.rvbsm.fsit.api.event;

import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.util.math.Vec3d;
import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import dev.rvbsm.fsit.entity.ModPose;

@FunctionalInterface
public interface UpdatePoseCallback {

    Event<UpdatePoseCallback> EVENT = EventFactory.createArrayBacked(
        UpdatePoseCallback.class, (listeners) -> (player, pose, pos) -> {
            for (UpdatePoseCallback listener : listeners) {
                listener.update(player, pose, pos);
            }
        });

    void update(@NotNull ServerPlayerEntity player, @NotNull ModPose pose, @Nullable Vec3d pos);
}

File: src/main/kotlin/dev/rvbsm/fsit/event/UpdatePoseCallback.kt
package dev.rvbsm.fsit.event

import dev.rvbsm.fsit.entity.PlayerPose
import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.util.math.Vec3d

fun interface UpdatePoseCallback {
    fun onUpdatePose(player: ServerPlayerEntity, pose: PlayerPose, pos: Vec3d?)

    companion object {
        @JvmField
        val EVENT: Event<UpdatePoseCallback> =
            EventFactory.createArrayBacked(UpdatePoseCallback::class.java) { listeners ->
                UpdatePoseCallback { player, pose, pos ->
                    for (listener in listeners) {
                        listener.onUpdatePose(player, pose, pos)
                    }
                }
            }
    }
}
